src/itk_torch_affine_matrix_bridge.py

```diff
@@ -0,0 +1,339 @@
+    inverse_direction_matrix = torch.eye(ndim+1, dtype=torch.float64) 
+    inverse_direction_matrix[:ndim, :ndim] = torch.tensor(inverse_direction, dtype=torch.float64)
+
+    return direction_matrix, inverse_direction_matrix
+
+def compute_reference_space_affine_matrix(image, ref_image): 
+    ndim = ref_image.ndim
+
+    # Spacing and direction as matrices
+    spacing_matrix, inv_spacing_matrix = [m[:ndim, :ndim].numpy() for m in compute_spacing_matrix(image)]
+    ref_spacing_matrix, ref_inv_spacing_matrix = [m[:ndim, :ndim].numpy() for m in compute_spacing_matrix(ref_image)]
+
+    direction_matrix, inv_direction_matrix = [m[:ndim, :ndim].numpy() for m in compute_direction_matrix(image)]
+    ref_direction_matrix, ref_inv_direction_matrix = [m[:ndim, :ndim].numpy() for m in compute_direction_matrix(ref_image)]
+
+    # Matrix calculation
+    matrix = ref_direction_matrix @ ref_spacing_matrix @ inv_spacing_matrix @ inv_direction_matrix 
+
+    # Offset calculation
+    pixel_offset = -1
+    image_size = np.asarray(ref_image.GetLargestPossibleRegion().GetSize(), np.float32)
+    translation =  (ref_direction_matrix @ ref_spacing_matrix - direction_matrix @ spacing_matrix) @ (image_size + pixel_offset) / 2 
+    translation += np.asarray(ref_image.GetOrigin()) -  np.asarray(image.GetOrigin())
+
+    # Convert matrix ITK matrix and translation to MONAI affine matrix
+    ref_affine_matrix = itk_to_monai_affine(image, matrix=matrix, translation=translation)
+
+    return ref_affine_matrix 
+
+
+def itk_to_monai_affine(image, matrix, translation, center_of_rotation=None, reference_image=None):
+    """
+    Converts an ITK affine matrix (2x2 for 2D or 3x3 for 3D matrix and translation
+    vector) to a MONAI affine matrix.
+    
+    Args:
+        image: The ITK image object. This is used to extract the spacing and 
+               direction information.
+        matrix: The 2x2 or 3x3 ITK affine matrix.
+        translation: The 2-element or 3-element ITK affine translation vector.
+        center_of_rotation: The center of rotation. If provided, the affine 
+                            matrix will be adjusted to account for the difference
+                            between the center of the image and the center of rotation.
+        reference_image: The coordinate space that matrix and translation were defined
+                         in respect to. If not supplied, the coordinate space of image
+                         is used.
+        
+    Returns:
+        A 4x4 MONAI affine matrix.
+    """
+
+    assert_itk_regions_match_array(image)
+    ndim = image.ndim
+
+    # If there is a reference image, compute an affine matrix that maps the image space to the
+    # reference image space.
+    if reference_image:
+        assert_itk_regions_match_array(reference_image)
+        assert image.shape == reference_image.shape, "ITK-MONAI bridge: shape mismatch between image and reference image"
```

Comment:
It seems strange to me that the reference image (which provides a reference coordinate system) and the image providing the index-to-physical space transform have to have the same shape.    Seems like all operations should be in terms of Index, Spacing, Origin, and Direction.   Line 129 uses image size...but image size is being multiplied by the direction matrix (which exists in physical space, not in index space).  That is, a direction matrix (which should consist only of unit vectors and should not do any scaling/spacing adjustments) shouldn't be multiplied by an image size (which exists in index space), unless image size is first multiplied by spacing, to bring it into physical space.